wrangler-service/src/main/java/co/cask/wrangler/service/common/AbstractWranglerService.java

```diff
@@ -71,4 +120,98 @@ protected boolean shouldCopyHeader(@Nullable String workspaceId) throws Workspac
+    }
+    
+    return ugiCache.get(impersonatedUser);
+  }
+  
+  private static UserGroupInformation createProxyUGI(String impersonatedUser)
+      throws IOException, InterruptedException {
+    if (impersonatedUser == null) {
+      return null;
+    }
+    
+    UserGroupInformation currnetUgi = UserGroupInformation.loginUserFromKeytabAndReturnUGI(
+        nsPrincipal, nsKeytab);
+    LOG.debug("current UGI {} , realUser {}, userShortName {} , userName {} , loginUser {}", 
+              currnetUgi, currnetUgi.getRealUser(), currnetUgi.getShortUserName(), currnetUgi.getUserName(),
+              currnetUgi.getLoginUser());
+      
+    UserGroupInformation proxyUgi = UserGroupInformation.createProxyUser(impersonatedUser, currnetUgi);
+    LOG.debug("proxyUgi UGI {} , realUser {}, userShortName {} , userName {} , loginUser {}", 
+              proxyUgi, proxyUgi.getRealUser(), proxyUgi.getShortUserName(), proxyUgi.getUserName(),
+              proxyUgi.getLoginUser());
+    return proxyUgi;
+  }
+
+  public String getLocalizedKeytabPath() {
+    if (!runtimeArgs.containsKey(PropertyIds.NAMESPACE_KETAB_PATH)) {
+      LOG.debug("Keytab config not found in Runtime arguments");
+      return null;
+    }
+      
+    String keytabPath = runtimeArgs.get(PropertyIds.NAMESPACE_KETAB_PATH);
+    Path p = Paths.get(keytabPath);
+    String file = p.getFileName().toString();
+    return file;
+  }
+  
+  protected boolean isUserImpersonationEnabled() {
+    if ((runtimeArgs.containsKey(PropertyIds.USER_IMPERSONATION_ENABLED)) &&
+              (runtimeArgs.get(PropertyIds.USER_IMPERSONATION_ENABLED).equalsIgnoreCase("true"))) {
+      return true;
+    }
+    return false;
+  }
+  
+  private static LoadingCache<String, UserGroupInformation> createUGICache() {
+    long expirationMillis = ugiCacheExpirationMs;
+    return CacheBuilder.newBuilder()
+      .expireAfterWrite(expirationMillis, TimeUnit.MILLISECONDS)
+      .build(new CacheLoader<String, UserGroupInformation>() {
+        @Override
+        public UserGroupInformation load(String key) throws Exception {
+          return createProxyUGI(key);
+        }
+      });
+  }
+
+  private void initialzeUGICache() {
+    if (ugiCache != null) {
+      return;
+    }
+    
+    synchronized (AbstractHttpServiceHandler.class) {
+      if (ugiCache != null) {
+        return;
+      }
+      ugiCacheExpirationMs = getUgiCacheExpirationMs();
+      ugiCache = createUGICache();
+    }
+  }
+  
+  private long getUgiCacheExpirationMs() {
+    Configuration conf = new Configuration();
+    long tokenRenewInterval = conf.getLong(DFSConfigKeys.DFS_NAMENODE_DELEGATION_TOKEN_RENEW_INTERVAL_KEY,
+        DFSConfigKeys.DFS_NAMENODE_DELEGATION_TOKEN_RENEW_INTERVAL_DEFAULT);
+    long retValue = DEFAULT_UGI_CACHE_EXPIRATION_MS;
+    if (tokenRenewInterval < DEFAULT_UGI_CACHE_EXPIRATION_MS) {
+      retValue = (tokenRenewInterval / 2);
+    }
+    LOG.debug("Setting cache expiration time to: {} ms", retValue);
```

Comment:
it would be better to use a constant gap between nn token review interval and ugi cache eviction period.

Reply:
As discussed, I have modified the logic now to use Hadoop config / 2 as expiration time